src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrProviderStateFactory.java

```diff
@@ -111,7 +111,20 @@ private static Bundle extractCallingBundle(@NotNull Map<String, Object> authenti
                     } else {
                         final Object subService = authenticationInfo.get(ResourceResolverFactory.SUBSERVICE);
                         final String subServiceName = subService instanceof String ? (String) subService : null;
-                        session = repo.loginService(subServiceName, null);
+                        // let's shortcut the impersonation for service users, if impersonation was requested
+                        String sudoUser = getSudoUser(authenticationInfo);
+                        if (sudoUser != null) {
+                            SimpleCredentials creds = new SimpleCredentials(sudoUser, new char[0]);
+                            creds.setAttribute(ResourceResolver.USER_IMPERSONATOR, subServiceName == null ?
```

Comment:
why is that attribute needed?
SlingRepository.impersonateFromService just takes a credential object and the method itself already implies that an impersonation is requested to be performed. 

the extra attribute is not needed afaik and the implementation of SlingRepository.impersonateFromService that is executed in AEM and the default in sling (https://github.com/apache/sling-org-apache-sling-jcr-base/blob/master/src/main/java/org/apache/sling/jcr/base/AbstractSlingRepository2.java#L391-L427) don't respect it. but maybe that explains my confusion with the comment (see above)

Reply:
That attribute seems to be retrieved via `org.apache.sling.jcr.resource.internal.helper.jcr.JcrResourceProvider#getAttribute` for `org.apache.sling.api.resource.ResourceResolver#getAttribute` calls.

Reply:
@raducotescu , as discussed in private.... ok with leaving it in since apparently the resource-resolver api allows to retrieve the impersonator and does so via a session attribute.
althought that feels a bit awkward to me to use a jcr-session attribute for that which by coincidence gets populated from session attributes *autsch*